### PR TITLE
nixos/systemd-boot: pass EFI variable flags during update too

### LINKED ISSUE #151336: systemd-boot update touches EFI variables when told not to
### Describe the bug
During systemd-boot update step `bootctl` attempts to touch EFI variables, even when `boot.loader.efi.canTouchEfiVariables` is set to `false`.

### Steps To Reproduce
1. Install NixOS 21.05 with systemd-boot and `boot.loader.efi.canTouchEfiVariables = false;`
2. Upgrade to NixOS 21.11 (fa8aa7ff5995d539aabc6f6a009dfcc1e3fd33a8)
3. Configuration switch fails with:
```
Copied "/nix/store/rgjrn2vl55pyxahfggvbdgl5g0xn0an1-systemd-249.5/lib/systemd/boot/efi/systemd-bootx64.efi" to "/boot/EFI/systemd/systemd-bootx64.efi".
Copied "/nix/store/rgjrn2vl55pyxahfggvbdgl5g0xn0an1-systemd-249.5/lib/systemd/boot/efi/systemd-bootx64.efi" to "/boot/EFI/BOOT/BOOTX64.EFI".
updating systemd-boot from (247) to (249.5)
Failed to create EFI Boot variable entry: Input/output error
Traceback (most recent call last):
  File "/nix/store/5f69p1syxlsfyxxb54zs1gc59ngf2cxz-systemd-boot", line 314, in <module>
    main()
  File "/nix/store/5f69p1syxlsfyxxb54zs1gc59ngf2cxz-systemd-boot", line 267, in main
    subprocess.check_call(["/nix/store/rgjrn2vl55pyxahfggvbdgl5g0xn0an1-systemd-249.5/bin/bootctl", "--path=/boot", "update"])
  File "/nix/store/wl02plhc6zf84m6x9984l42wnnnbly5m-python3-3.9.6/lib/python3.9/subprocess.py", line 373, in check_call
    raise CalledProcessError(retcode, cmd)
subprocess.CalledProcessError: Command '['/nix/store/rgjrn2vl55pyxahfggvbdgl5g0xn0an1-systemd-249.5/bin/bootctl', '--path=/boot', 'update']' returned non-zero exit status 1.
```
An error is just an indication of update trying to touch EFI variables. (I have EFI vars touching disabled on that system because of some unrelated firmware breakage)
Manually running `bootctl --path=/boot update --no-variables` properly finishes, further switches work fine.

### Expected behavior
systemd-boot update should not touch EFI variables during upgrade.

### Notify maintainers
@andir @flokli @kloenk

### Metadata
Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the result.

```console
[user@system:~]$ nix-shell -p nix-info --run "nix-info -m"
 - system: `"x86_64-linux"`
 - host os: `Linux 5.10.64, NixOS, 21.11 (Porcupine)`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.3.16`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`

```

Maintainer information:
```yaml
# a list of nixpkgs attributes affected by the problem
attribute:
# a list of nixos modules affected by the problem
module:
```

### PR DESCRIPTION
**Pull request by tpwrules**
On some systems, EFI variables are not supported or otherwise wonky. bootctl attempting to access them causes failures during bootloader installations and updates. For such systems, NixOS provides the options `boot.loader.efi.canTouchEfiVariables` and `boot.loader.systemd-boot.graceful` which pass flags to bootctl that change whether and how EFI variables are accessed.

Previously, these flags were only passed to bootctl during an install operation. However, they also apply during an update operation, which can cause the same sorts of errors. This change passes the flags during update operations as well to prevent those errors.

Fixes https://github.com/NixOS/nixpkgs/issues/151336

###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [X] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [X] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: nixos/modules/system/boot/loader/systemd-boot/systemd-boot-builder.py [hunk 1]
```diff
@@ -228,20 +228,21 @@ def main() -> None:
         warnings.warn("NIXOS_INSTALL_GRUB env var deprecated, use NIXOS_INSTALL_BOOTLOADER", DeprecationWarning)
         os.environ["NIXOS_INSTALL_BOOTLOADER"] = "1"
 
+    # flags to pass to bootctl install/update
```
Comment (winterqt) [inline nixos/modules/system/boot/loader/systemd-boot/systemd-boot-builder.py:231]: This comment feels redundant, maybe? The name is kind of obvious...

**Review by winterqt — APPROVED** [review]
I doubt we can really test for this behavior in the existing systemd-boot update test, right?

(Also: I think we should backport this. (Sure, it's technically a breaking change, but it would only be fixing something that was already broken.))

**Comment by winterqt** [comment]
@ofborg test systemd-boot

File: nixos/modules/system/boot/loader/systemd-boot/systemd-boot-builder.py (hunk 1 above)
Comment (tpwrules) [reply nixos/modules/system/boot/loader/systemd-boot/systemd-boot-builder.py:231]: You know I actually went back and forth for a while on the comment and variable name here. My defense is that the flags aren't passed to the bootctl version and status checks.

Comment (winterqt) [reply nixos/modules/system/boot/loader/systemd-boot/systemd-boot-builder.py:231]: Ooh, good point. Wonder if there's a better name for it... `bootctl_install_flags` could work, but that would also require a comment explaining that it's passed to both. But I also think `bootctl_flags` is a tiny bit disingenuous? I don't know, maybe I'm nitpicking.

Comment (tpwrules) [reply nixos/modules/system/boot/loader/systemd-boot/systemd-boot-builder.py:231]: And `flags` was kind of ambiguous now that the scope has changed and `install_and_update_flags` is annoyingly long and...

I am pretty sure there's no obvious win here and we should keep it how it is :)

**Comment by winterqt** [comment]
```
Traceback (most recent call last):
  File "/nix/store/d5s1pdjh8vkr9id10jbs1fc5fdy3cy6l-nixos-test-driver-1.1/bin/.nixos-test-driver-wrapped", line 9, in <module>
    sys.exit(main())
  File "/nix/store/d5s1pdjh8vkr9id10jbs1fc5fdy3cy6l-nixos-test-driver-1.1/lib/python3.10/site-packages/test_driver/__init__.py", line 112, in main
    driver.run_tests()
  File "/nix/store/d5s1pdjh8vkr9id10jbs1fc5fdy3cy6l-nixos-test-driver-1.1/lib/python3.10/site-packages/test_driver/driver.py", line 142, in run_tests
    self.test_script()
  File "/nix/store/d5s1pdjh8vkr9id10jbs1fc5fdy3cy6l-nixos-test-driver-1.1/lib/python3.10/site-packages/test_driver/driver.py", line 138, in test_script
    exec(self.tests, symbols, None)
  File "<string>", line 12, in <module>
AssertionError
```

(per the test run)

**Comment by winterqt** [comment]
Ah, here we go: https://github.com/NixOS/nixpkgs/blob/ea27fb07efe68cb426593a4d7c2efe469621021d/nixos/tests/systemd-boot.nix#L110

**Comment by tpwrules** [comment]
It might be possible to test if qemu can simulate a read only variable store.

I don't see how the test failure could be related to my change? How does one even view ofborg's output anyway?

**Comment by winterqt** [comment]
> How does one even view ofborg's output anyway?

image

**Comment by winterqt** [comment]
> I don't see how the test failure could be related to my change?

Yeah, I don't either, since the update test's configuration just passes the same flags to `bootctl update` as it always would...

**Comment by winterqt** [comment]
Ah, yup, it's been failing for a bit now: https://hydra.nixos.org/job/nixos/trunk-combined/nixos.tests.allDrivers.systemd-boot.update.aarch64-linux

**Comment by tpwrules** [comment]
I think we just need to increase the size of the FS in the test? The kernel is uncompressed on aarch64-linux so it's rather large. I think it plus the initrd must have just recently eclipsed 60M.

**Comment by winterqt** [comment]
That's not the issue I'm talking about, sorry -- the x86_64 issue is different, and it's currently passing on Hydra.

**Comment by tpwrules** [comment]
You had the wrong job, the actual test has definitely been [failing for a while](https://hydra.nixos.org/build/205381115).

Root cause appears to be that systemd-boot and/or our update script started printing `updating systemd-boot from 000.0-1-notnixos to` while the test script is expecting `updating systemd-boot from (000.0-1-notnixos) to`.

There are also other deficiencies including disk space errors on aarch64-linux and some reproducibility problems. Might try to revamp this into a systemd-boot refresh PR. But I still stand by it as it is now.

**Comment by tpwrules** [comment]
Superseded by https://github.com/NixOS/nixpkgs/pull/211723

**Pull request closed without merging**